Fix issues encountered during pytest

michael-okeefe opened the pull request:
Fixes some issues encountered during running `pytest -v` on Windows:

* updates how Python is called for demo tests so that it works within a virtual environment
* Handles how Polars DataFrame is printed by first transforming to a Pandas DataFrame

Additionally, this version adds a CI workflow file to build and test the application.

Commit:
Merge branch 'fastsim-3' into f3/fix-pytest-issues

Commit:
Add source dist and wheel builds; save artifacts.

This updates the `build_and_test.yaml` file to also
build source distributions (sdist), wheel files, and
to also save artifacts of those builds.

Commit:
Fix typo with quotes

Commit:
Fix issue with string replacement.

We updated the python versions to be the full version
(e.g., 3.8, 3.9, or 3.10) versus just the minor version
(e.g., 8, 9, 10). However, the string replacement code
was not updated appropriately that got copied over for
cibuildwheel. This has been corrected.

Commit:
Go back to using just python minor version.

This makes the string manipulation easier for cibuildwheel.

Commit:
Update build_and_test file to include pytest for test_speedup.py

Commit:
Disable speed tests without environment variable

Commit:
Update for both x86 as well as apple silicon targets

Commit:
Turn speedup tests off by default

Commit:
Conditionally add target architecture

Commit:
Add required path to tests for cibuildwheel to work

Commit:
Revert to unittest suite (vs pytest)

Commit:
Iterating on build settings for cibuildwheel

Commit:
Combine target names and set artifact names.

Commit:
Temporarily reduce matrix size

Commit:
Reduce run matrix

Commit:
Improve artifact names; try macos CI

Commit:
Correct syntax error in YAML

Commit:
Adjust settings to get CI to build MacOS X

Commit:
Add back MacOS arm64

Commit:
Switch to arm64

Commit:
Set up Python to use native python for macos (Apple Silicon)

Commit:
Try reverting cibuildwheel command to FS2

Commit:
Update workflow to try to get macos to build

Commit:
Differentiate artifact names

Commit:
Run with full matrix

Commit:
Revert to closer to fastsim-2 workflow

Commit:
Switch to new Github Action for rust installation

Commit:
Correct matrix variable path

Commit:
Add os version to build string

Commit:
Adjust macos build configs for rust

Commit:
Run only macos-14 for testing; change version from number to string

Commit:
Add x86_64 for MacOS-14

Commit:
Explicitly set MacOS arch vs universal2.

Note: this is to get around issues encountered on CI
using delocate-wheel which is failing.

Commit:
Include macos-13

Commit:
Re-enable all os/python versions

Commit:
Update Python versions to test

Commit:
Update python versions to include 3.11

Commit:
Bump min version of python to range tested.

Python 3.10 and 3.11 are being tested on the CI.
This bumps the minimum range up to 3.10 from 3.8.

Note: as of today, 3.8 is end-of-life.

michael-okeefe committed:
Set minimum MacOS deployment target for Intel mac builds

Try setting the MACOS_DEPLOYMENT_TARGET equal to 10.12 per instructions from failed pipeline.

Commit:
Merge branch 'fastsim-3' into f3/fix-pytest-issues

Commit:
Merge branch 'fastsim-3' of github.com:NREL/fastsim into f3/fix-pytest-issues

calbaker commented on .github/workflows/build_and_test.yaml:
```diff
@@ -4,34 +4,108 @@ on: [push]
+          - base: ubuntu
+            version: latest
+          - base: macos
+            version: '13'
```
@michael-okeefe , could you drop some comments in here explaining why `latest` does not work for `macos` and why `13` and `14` are used here?  I want to make sure this is not hard to decipher in the future.

calbaker replied:
@michael-okeefe , please also explain in the comments what event should trigger adding `15`, `16`, ...

michael-okeefe replied:
Sure, I can do that. The issue is related to what processor each github CI machine is running for mac. MacOS 13 is Intel and MacOS 14 is Apple Silicon. Ideally, the Apple transition to their new chip occurs fast and we can drop having to call out specific machines and then just use latest, etc. 15, 16, etc. can be brought in at any time we desire and they should look like the calls for 14. As soon as 13 drops from the board, we can go back to just saying 'latest'.

michael-okeefe replied:
@calbaker Note added to clarify why we are calling out macos-13 and 14.

calbaker commented:
> Handles how Polars DataFrame is printed by first transforming to a Pandas DataFrame
@michael-okeefe , where did this change happen?  I'm struggling to find it.

michael-okeefe commented:
> > Handles how Polars DataFrame is printed by first transforming to a Pandas DataFrame
> > @michael-okeefe , where did this change happen?  I'm struggling to find it.

@calbaker Apologies, I must have gotten my PRs mixed up as I don't see the change here, either. This was the area where the issue occurred during tests:

https://github.com/NREL/fastsim/blob/f3/fix-pytest-issues/python/fastsim/demos/demo_variable_paths.py#L49

Not sure if this is the exact line or not, though.

michael-okeefe committed:
Add note for os matrix

The Mac OS options might be hard for others to follow. Here, we add a note to help others following behind to understand what we're doing and why.

calbaker commented on .github/workflows/build_and_test.yaml:
```diff
@@ -7,6 +7,9 @@ jobs:
       fail-fast: true
+      # NOTE: In the matrix below, we call out macos 13 and 14 explicitly. This is to enable running on both Intel and Apple Silicon for Mac.
+      # As of this writing, MacOS 13 is Intel and MacOS 14 is Apple Silicon. Ideally, the Apple transition to their new chip occurs fast and
+      # we can drop macos-13 checks and then just use 'latest'.
```
@michael-okeefe , could we use `13` and `latest` so that we don't have to remember to increment the `14`?

michael-okeefe replied:
@calbaker Great idea. I made the change and the tests are running.

michael-okeefe committed:
Switch to use latest vs 14 for macos
